### ISSUE TITLE
Implement Precompile Shader Archive

### ISSUE THREAD

**Issue by ptitSeb** [ptitSeb/gl4es#117]

**Comment by ptitSeb** [comment]
Implemented in commit 601184b1b7e63cd1f85b5360207bb763efa8fe82

**Comment by kas1e** [comment]
While Daniel working on adding necessary functions to amigaos4 driver, i still did test how it all handles currently by settings LIBGL_NOPSA to 0. 

So, i just run neverball and exit. As result, it create .gl4es.psa file in the root of game (good!), which is 43 bytes of size  (so just header saying GL4ES PrecompiledShaderArchive + some bytes of structure format or something).  When i exit from game i also have at end "LIBGL: Saved a PSA with 0 Precompiled Programs" (that of course expected, as there is no functions implemented).

Through, should't there be error or something (or log message) saying when running something like "LIBGL: Forcing to use PSA, but functions didn't work" , or something ? Or it make no sense ?

**Comment by ptitSeb** [comment]
Mmm, yes, the archive should not be created. I'll check that !

**Comment by ptitSeb** [comment]
@kas1e : with comit 37d3b629580f2289703d5ee83f701c38cabf08ef it should now ignore PSA if the extension is not present (you will have to delete yourself the previous empty one).

**Comment by kas1e** [comment]
Btw, did ypu test PSA on Pandora already ? I.e. with fricking shark while playing and loading of neverball for example. Interesting to know if on your side with pandora's gles2 all works as expected now ?

**Comment by ptitSeb** [comment]
I tested quickly on the Pandora, and it seems promising yes. Starting time for Neverball is much faster now. I need to do more testing, but that looks good (I haven't tried Friking Shark yet).

**Comment by kas1e** [comment]
Is pandora's gles2 driver also have needs to assembly from spirv format those precompiled shaders, or they already in the machine code and not in spirv saved in .gl4es.psa file ?

**Comment by ptitSeb** [comment]
No spirv on the pandora, it's already machine code specific to the PowerVR it use.

**Comment by kas1e** [comment]
So if then take in account that Hans will not do it in Nova, then maybe asking Daniel to save psa not in spirv, but in Nova's assembly code will make sense.. (If nova provide such conversion functions to public)

**Comment by ptitSeb** [comment]
yep

**Comment by ptitSeb** [comment]
But note that I don't think it's possible to access Nova compiled program shader if Hans don't write the equivalent of `glProgramBinary(...)` and `glGetProgramBinary(...)` into Nova. That's why Daniel proposed to do it using the SpirV way.

**Comment by ptitSeb** [comment]
Yes, the .gl4es.psa is supposed to be build on the target computer, not bring by the app. That means the 1st run will still get the hicup, but other runs will not. It the same on the Pandora, there are 3 different model with slightly diferent hardware, and there are a lot of version of the driver user can choose from, so I don't plan to put any pre-baked gl4es.psa in games, and will one be build by itself.

**Comment by kas1e** [comment]
Aha got it.. So let's wait and see if implementation as Daniel do with "spirv" on top of nova conversion will give enough boost or not. If it will be enough, then our version will be even portable across different gfx card, but if not , then probabaly we will need ask Hans to provide public api to generate machine ready code.

**Comment by kas1e** [comment]
Hi :)

So Daniel send me first version with his implementation of necessary functions, that what readme says:

- Implemented the OES_get_program_binary extension as requested by Kasie Kasovich (kas1e). Because Nova doesn't provide any support here, this implementation works with the internal SPIR-V code of all shaders that make up the respective program. So the binaries here are no true machine-code but intermediate code, packed into a proprietary format. Providing such binaries essentially skips the vertex- and fragment-shader's GLSL compilation steps. Anyway, this means:
 - new function GetProgramBinaryOES to extract such a binary representation of the respective successfully linked shader program.
- new function ProgramBinaryOES to supply the GL with such a cached binary.
- support for the glGet parameter GL_NUM_PROGRAM_BINARY_FORMATS_OES - which returns 1 now.
- support for the glGet parameter GL_PROGRAM_BINARY_FORMATS_OES, which returns one single format ID identifying my binary format (0xC00DA675)
- support for the glGetProgramiv parameter GL_PROGRAM_BINARY_LENGTH_OES which gives you the buffer size required to store the binary for the respective program.
- consequently add GL_OES_get_program_binary to the extension string.	


So, i run Neverball over new binary, and gl4es says me in output:

LIBGL: Extension GL_OES_get_program detected
LIBGL: Numberof supported Program Binary Format : 1

And nothing more related to it.

So , after i exit from Neverball, i didn't have any .gl4es.psa file created. I run some DOS tracer (to see if it tries to create that file) and nope, file even didn't tries to be created.

Have any clue what to check next ?:)

**Comment by ptitSeb** [comment]
Do you see "LIBGL: Shuting down" at the end of the program at least?

The function `close_gl4es()` from `src/gl/init.c` at this end of the file, declared as a "destructor" should print it.

**Comment by kas1e** [comment]
Yeah of course i see that one, i just print only info relevant to PSA, but there is full log (supertuxkart0.6.2a, latest gl4es):

```
4/0.Work:games/supertuxkart> supertuxkart_gl4es_1916 
LIBGL: Initialising gl4es
LIBGL: v1.1.1 built on Jul 22 2019 23:35:10
LIBGL: Using GLES 2.0 backend
LIBGL: Using Warp3DNova.library v1 revision 65
LIBGL: Using OGLES2.library v2 revision 9
LIBGL: OGLES2 Library and Interface open successfuly
LIBGL: Targeting OpenGL 2.0
LIBGL: Forcing NPOT support by disabling MIPMAP support for NPOT textures 
LIBGL: Not trying to batch small subsequent glDrawXXXX
LIBGL: Current folder is:/Work/games/supertuxkart
Data files will be fetched from: '.'
Highscores will be saved in './.supertuxkart/highscore.data'.
LIBGL: Hardware test on current Context...
LIBGL: Hardware Limited NPOT detected and used
LIBGL: Extension GL_EXT_blend_minmax detected and used
LIBGL: FBO are in core, and so used
LIBGL: PointSprite are in core, and so used
LIBGL: CubeMap are in core, and so used
LIBGL: BlendColor is in core, and so used
LIBGL: Blend Substract is in core, and so used
LIBGL: Blend Function and Equation Separation is in core, and so used
LIBGL: Texture Mirrored Repeat is in core, and so used
LIBGL: Extension GL_OES_mapbuffer detected
LIBGL: Extension GL_OES_element_index_uint detected and used
LIBGL: Extension GL_OES_packed_depth_stencil detected and used
LIBGL: Extension GL_EXT_texture_format_BGRA8888 detected and used
LIBGL: Extension GL_OES_texture_float detected and used
LIBGL: high precision float in fragment shader available and used
LIBGL: Extension GL_EXT_frag_depth detected and used
LIBGL: Max vertex attrib: 16
LIBGL: Extension GL_OES_get_program detected
LIBGL: Number of supported Program Binary Format: 1
LIBGL: Max texture size: 16384
LIBGL: Max Varying Vector: 32
LIBGL: Texture Units: 8(8), Max lights: 8, Max planes: 6
LIBGL: Extension GL_EXT_texture_filter_anisotropic detected and used
LIBGL: Max Anisotropic filtering: 16
LIBGL: Hardware vendor is A-EON Technology Ltd. Written by Daniel 'Daytonta675x' MьЯener @ GoldenCode.eu
LIBGL: OGLES2 Library and Interface closed
LIBGL: Shuting down 
```

**Comment by kas1e** [comment]
I also asking Daniel to recheck new stuff too, he says that glGetProgramiv with GL_PROGRAM_BINARY_LENGTH_OES works and glGetProgramBinaryOES apparently works too.

**Comment by kas1e** [comment]
Btw, measuring things by some test examples on os4 for now, we found that of course all the cached binaries should be preloaded right on running, because loading binaries from disk can slow the things down the same as before. 

But i assume in case with gl4es, on begining it scaning for .gl4es.psa , and if it found and functions/extensions working, it then preload binaries to memory, right ?

**Comment by ptitSeb** [comment]
Check in https://github.com/ptitSeb/gl4es/blob/master/src/gl/init.c#L569 if the name of the PSA is correct, using some `printf`

**Comment by ptitSeb** [comment]
> Btw, measuring things by some test examples on os4 for now, we found that of course all the cached binaries should be preloaded right on running, because loading binaries from disk can slow the things down the same as before.
> 
> But i assume in case with gl4es, on begining it scaning for .gl4es.psa , and if it found and functions/extensions working, it then preload binaries to memory, right ?

Yes. Loading of the PSA file is done at init of gl4es. Then, when a fpe shader need to be created, it ceck first in the PSA archive (in memory), and create the shader directly if present. If not present, the shader is created in the traditional way, and is added to the in-memory PSA (that is then flaged "dirty"). At shut down of gl4es, the PSA archive is writen back to disk if it's flagged dirty.

**Comment by kas1e** [comment]
> Check in https://github.com/ptitSeb/gl4es/blob/master/src/gl/init.c#L569 if the name of the PSA is correct, using some printf

Something wrong with whole "If" started from if(hardext.prgbin_n>0) { line. Maybe missing } or something somewhere, because none printfs is reached there.

As i see for now , all that file creation thing inside of if(globals4es.nopsa==0) { } loops, which probabaly not what it should be ?

**Comment by ptitSeb** [comment]
look closely, it's *no*psa , so it should be 0 to use psa.

**Comment by kas1e** [comment]
That how it looks like for me now:

```
    if (getcwd(cwd, sizeof(cwd))!= NULL)
        SHUT(LOGD("LIBGL: Current folder is:%s\n", cwd));

	printf("hardext.prgbin_n = %d\n", hardext.prgbin_n);

	printf("uuuuuuuuuuuuuu\n");	
	SHUT(LOGD("OMEGA: 2232323232323ooooo\n"));

    if(hardext.prgbin_n>0) {
        env(LIBGL_NOPSA, globals4es.nopsa, "Don't use PrecompiledShaderArchive");
		printf("eeeeeeeeee\n");
		
        if(globals4es.nopsa==0) {
			printf("erewrerrrrrr\n");
            cwd[0]='\0';
            // TODO: What to do on ANDROID and EMSCRIPTEN?
#ifdef __linux__
            const char* home = getenv("HOME");
            if(home)
                strcpy(cwd, home);
            if(cwd[strlen(cwd)]!='/')
                strcat(cwd, "/");
#elif defined AMIGAOS4
            strcpy(cwd, "PROGDIR:");
#endif

			SHUT(LOGD("OMEGA: ooooooooooooo\n"));
            if(strlen(cwd)) {
				SHUT(LOGD("OMEGA: bbbbbbbbbbbbbb\n"));
                strcat(cwd, ".gl4es.psa");
				printf("cwd = %s\n", cwd);
                fpe_InitPSA(cwd);
                fpe_readPSA();
            }
        }
    }
}
```

And in log, i can see :

hardtext.prgbin_n = 0
uuuuuuuuuuuuuuuuu
OMEGA:22232323232323oooooo

and no other prinfs

**Comment by ptitSeb** [comment]
Mmmm, wait, on AmigaOS, there is no pre-tests, so the prog_n is 0 by default.
What is the version of GLES2 that have this extension available ?

**Comment by kas1e** [comment]
that one where those new functions added ? 2.9

**Comment by kas1e** [comment]
Or you mean GLES2 standard ?

**Comment by ptitSeb** [comment]
> that one where those new functions added ? 2.9

Yes, that. Because I'll probably need to hardcode prog_n = 1 on AMIGA if OGLES2 driver >= 2.9 for now
(until I find a way to create an offscreen context on Amiga so I can launch some test at start of gl4es).

**Comment by kas1e** [comment]
Yeah, let's do it that way.
Or , you simple can assume that ogles2.9 is minimum for gl4es at all

**Comment by ptitSeb** [comment]
Mmm, wait, I have this in `src/agl/agl.c` I have
```
#define MIN_OGLES2_LIB_VERSION 1
#define MIN_OGLES2_LIB_REVISION 22
```
So, 1.9 doesn't work...

**Comment by kas1e** [comment]
current one are 2.9, so we can safely cahnge it to 2.9 in agl.c

**Comment by ptitSeb** [comment]
Ah ok. I'll do that.

**Comment by kas1e** [comment]
Probably it also enough in init.c, right before if(hardext.prgbin_n>0) { , add something like:

```
#ifdef __amigaos4__
hardext.prgbin_n=1
#endif

if(hardext.prgbin_n>0) {
```

?

**Comment by ptitSeb** [comment]
I have made the change to enabled. Now maybe something is not working correctly in gl4es / ogles2.
Like the `aglGetProcAddress(...)` ?

**Comment by ptitSeb** [comment]
Like, in `src/agl/amiga.c` change line 802 from
`return IOGLES2->aglGetProcAddress(name);`
to
```
void * ret = IOGLES2->aglGetProcAddress(name);
printf("aglGetProcAddress(%d) => %p\n", ret, name);
return ret;
```

**Comment by kas1e** [comment]
Just #include "hardext.h" fail to find the file, so i change to #include "../glx/hardext.h" , and then builds fine.

But then by some reassons this line:

```
if (!(LOGLES2->lib_Version > 2 || (LOGLES2->lib_Version == 2 && LOGLES2->lib_Revision >= 9)))  {
```

Did't meet the requements. But version of library in use is cleary 2.9.

Then, even if i remove that line, and hardcore put just 

hardext.prgbin_n = 1;

then nothing happens : i.e. no crash, but no file creates on exit. Like, its still again thin that hardext.prgbin is 0

**Comment by kas1e** [comment]
Requements fail because (!( , should be without ! there

**Comment by kas1e** [comment]
Its like "hardext.prgbin_n = 1;" doing in amigaos.c didn't enough and seems overwriten later by 0 again ?

**Comment by ptitSeb** [comment]
Yes, this should all be fixed now.

**Comment by kas1e** [comment]
Ok downloaded all from scratch of latest vesion :  now hardext.prgbin_n  1 for sure, but then it crashes as before when i put the same hardext.prgbin_n = 1 in other place.

So i add what you say and it print that:

aglGetProcAddress(glGetProgramBinaryOES) => 0x7f45d264

I also tried to use debug version of new ogles2.library (for better stack trace), and result that is crashes inside of OGLES2's glGetProgramBinaryOES()

**Comment by ptitSeb** [comment]
So you see with Daniel?

**Comment by kas1e** [comment]
You mean its on our side for sure ?

**Comment by ptitSeb** [comment]
Yes this function will be called. It's the function used to retreive a Shader to be stored in the PSA archive.
It works fine on the Pandora, and the function is not super complicated.

Look in `gl4es_getProgramBinary(...)` in `src/gl/program.c`

It works in two part: first the `gles_glGetProgramiv(glprogram->id, GL_PROGRAM_BINARY_LENGTH_OES, &l);` to get the size of the program (you should `printf("Sizeof program = %d\n", l);`) and then the actual `gles_glGetProgramBinary(glprogram->id, l, length, format, *binary);` to retreive the program...

**Comment by kas1e** [comment]
Ok, a wrote to Daniel about.. Did i understand right, that we use aglGetProcAddress() actually in gl4es only one time : exactly for glGetProgramBinaryOES() , or, it used all the time with other functions and works (so problem is them about glGetProgramBinaryOES() ? )

**Comment by ptitSeb** [comment]
It's the first time we used `aglGetProcAddress(...)` indeed, so it may be wrong.

You can test be commenting line 668 ( that is `EX(glClear)`) to force using `aglGetProcAddress(...)` also  for `glClear(...)` to be sure.

**Comment by kas1e** [comment]
Yeah crashes in glClear now

**Comment by kas1e** [comment]
Interesting .. Maybe something in our amiga related code broken when we return not just NULL ?

**Comment by ptitSeb** [comment]
I don't know, maybe it's the call convention? I have no idea, and no Amiga to experiment. You should talk about that with Daniel.

One thing I have noted and is quite surprising, is that I remember that I cannot simply used the address of `IOGLES2->glClear`, but needed to wrap is in `static void AmiglClear(...)` and use that function address. So I'm not sure what address `aglGetProcAddress(...)` is sending, but it should not be the IOGLES2 one (that I think have more hidden parameters), but the internal one...

**Comment by kas1e** [comment]
Maybe that because of needs to pass parameters as well ? 

I see for glColor we have before:

```
static void AmiglClear (GLbitfield mask) {
    return IOGLES2->glClear(mask);
}
```

Then EX(glClear) , which (and that define/redefine mess i don't understand for sure):

```
// Using glXXX name, return the function pointer of that function in ogles2 library
#define MAP(func_name, func) \
    if (strcmp(name, func_name) == 0) return (void *)Ami##func;

#define EX(func_name) MAP(#func_name, func_name)
```

Strange ... Maybe it needs ** , not just *

**Comment by ptitSeb** [comment]
That `EX` is a simple macro, and the point of the whole `aglGetProcAddress(...)` is to get a function pointer address for the function name given in parameter...
So calling `aglGetProcAddress("glClear")` will give the address of `static void AmiglClear (GLbitfield mask)` that just call `return IOGLES2->glClear(mask)`. No issue here, this is normal.

**Comment by kas1e** [comment]
I also checked the readme, and it says:

```
void* aglGetProcAddress(const char* name);

Returns the function pointer for the OGLES2 function with the respective name.
Returns nullptr if no such function exists.
```

So also looks pretty sane..

**Comment by ptitSeb** [comment]
Still, if it crash when gl4es tries to use it, there is something wrong...

**Comment by ptitSeb** [comment]
my question is, does this `glClear(GLbitfield mask)` function prototype, in OGLES2 really is that, or something that `glClear(something_t OGLES2, GLbitfield mask)` ?

**Comment by kas1e** [comment]
While Daniel seems sleep, i tried to check includes, and that what i find in 2 places:

#define glClear(mask) IOGLES2->glClear((mask)) 
and
void APICALL (*glClear)(struct OGLES2IFace *Self, GLbitfield mask); 

but not sure if it can answer on your question of course.. need to wait Daniel's answer

**Comment by kas1e** [comment]
I may try to create some simple test programm maybe.. Something like pure SDL2/GL app, which just call glClear(), but not directly, but from aglGetProcAddress, etc ..

**Comment by ptitSeb** [comment]
Ah yes, that: `void APICALL (*glClear)(struct OGLES2IFace *Self, GLbitfield mask);` define the function pointer for `glClear` with 2 paramters !, the first one beeing `OGLES2` itself.
If the `aglGetProcAddress` gives that, then indeed it will crash, because OGLES2 will never get passed (as it's not in the regular `glClear(...)` convention, and the program doesn't even have the OGLES2 structure itself.

**Comment by ptitSeb** [comment]
To test, define the function pointer, with something like
```
typedef void (*GLCLEARPTR)(GLbitfield mask);
```
(if you don't have `GLbitfield`, a simple `int` will be enough)
then in your main, after the GLES2 context is created
```
GLCLEARPTR my_glClear = (GLCLEARPTR)aglGetProcAddress("glClear");
if(my_glClear) {
 printf("Go!\n");
 my_glClear(GL_COLOR_BUFFER_BIT);
} else
 printf("glClear is NULL!\n");
```

**Comment by kas1e** [comment]
Do you mean comment out just EX(glClear) in gl4es, and all other stuff doing this inside of the programm's body, not in gl4es ?

**Comment by ptitSeb** [comment]
Nope, that just for your sample program using SDL2 / pure GLES2

**Comment by kas1e** [comment]
It say "Go!" and then crashes the same in glClear ..

**Comment by ptitSeb** [comment]
And now, try this changes:
`typedef void (*GLCLEARPTR)(struct OGLES2IFace *, GLbitfield);`
and
```
GLCLEARPTR my_glClear = (GLCLEARPTR)aglGetProcAddress("glClear");
if(my_glClear) {
 printf("Go!\n");
 my_glClear(IOGLES2, GL_COLOR_BUFFER_BIT);
} else
 printf("glClear is NULL!\n");
```

**Comment by kas1e** [comment]
Now also saying "Go!" , but this time no crash ! And i can exit from app without problems

**Comment by ptitSeb** [comment]
So I think you have enough information to open a bug to Daniel: for gl4es, I need the first one to work, not the second.

Now, if the first alternative (the pointer function to `glClear(GLbitfield)`) is not possible, I can think of an alternative, but that would be cumbersome, so I would prefer not to do it.

**Comment by kas1e** [comment]
Maybe it the way how amiga libraries works .. not sure he do anything special there, but will ask..

**Comment by kas1e** [comment]
Probabaly because of that reasson we have those static amiga functions as well..

**Comment by kas1e** [comment]
But is that not possible easy typedef things somehow, so first one will be just skipped, and second one will be taken ? I just not sure what exactly i need ask from Daniel to do ..

**Comment by ptitSeb** [comment]
You need to tell Daniel that the function pointer returned by `aglGetProcAddress(...)` for `glClear(...)` for example, is not for `void glClear(GLbitfield)` as it's supposed to be, but for `void glClear(struct OGLES2IFace *, GLbitfield)`

**Comment by kas1e** [comment]
Send a mail to Daniel with describing everythign we do (with our tests and co) , cross the fingers that it can be fixed/changed. It can be easy that this just the way how our amiga libraries works. Because if in readme says "Returns the function pointer for the OGLES2 function with the respective name." , then or it a bug on implementation side, or dunno what.. Waiting for answer from him :)

**Comment by kas1e** [comment]
Ok got an answer:

Okay, I get it. He's using GetProcAddr, casts the returned value to e.g. PFNGLGETPROGRAMBINARYOESPROC - which unfortunately doesnt take into account the "hidden" Amiga interface pointer.

So the solution is that I change my glGetProcAddr to not return the "normal" function pointer but instead I have to wrap everything into a dummy function without the struct pointer parameter, which I have to store as simple "global" (should be enough here).

Sure, he could work around by casting it to the correct type and by invoing it in the right way. But yes, it *is* cumbersome and doing it the way I outlined is the way to go for compatibility.


In other words Daniel will deal with, waiting for new beta :)

**Comment by ptitSeb** [comment]
Ah good :)

**Comment by kas1e** [comment]
Got fixed version : all start to works. .gl4es.psa created, etc.

Through at moment with Neverball it change nothing in terms of first loading : when i don't have .gl4es.psa and start the game and if i have .gl4es.psa and start the game : all the same by speed :(

Then i tried fricking shark (the most noticable by issues game) : also no changes.

I mean, absolutly no changes :) Even on single bit , which is strange ... Need invistigate a bit more.

I also notice that Neverball loose 20 fps by some reassons, but not sure if it new olges2, or latest gl4es, need to recheck..

**Comment by ptitSeb** [comment]
On the Pandora, on Neverball and Foobillard and I a very noticable change on 2nd start yes. There was a risk that the "SPIR-V" trick used for the ProgramBinary would not make things faster... That may be that indeed.
Everything happens inside `gl4es_useProgramBinary(...)` if you want to check (you can try to put some printf log with some timestaps if you want to measures stuffs).

Also, you lost 20fps in Niverball? Strange, I haven't noticed any loss in speed lately in my testing (and I tested latest Neverball this week end). But you can do `git bisect` or just take an old build to see if it was better before. I'm interested if I made some "slowing down" changes.

**Comment by kas1e** [comment]
About neverball loosing the speed: sorry it was false alarm, its just i in process of testing enable VSYNC :)

Now to PSA:

Can you point out plz on the source where PSA is generated ? 

Daniel checked one of generated files, and says that something is wrong (unless there some weird file format) : The first 8 bytes of every of my binary prog seems to be repeated. That's certainly not coming from me. So if my binary's 8 starting bytes would look like C0 0D A6 75 (4 bytes header) 00 00 1E 67 (4 bytes sizeof binary),  then inside gl4es psa file this sequence is not followed by the actual data but instead this 8 bytes are repeated. THEN the data starts. But this is definitely not what I return.

The binary in question is our .gl4es.psa generated when running Neverball:
http://kas1e.mikendezign.com/aos4/gl4es/PSA/gl4es_psa_neverball

**Comment by ptitSeb** [comment]
It's in multiple piece, but:

A new shader is added in memory here: https://github.com/ptitSeb/gl4es/blob/master/src/gl/fpe_cache.c#L262
A shader itself is read from GLES2 here: https://github.com/ptitSeb/gl4es/blob/master/src/gl/program.c#L685

And it's writen to disk here: https://github.com/ptitSeb/gl4es/blob/master/src/gl/fpe_cache.c#L173

**Comment by kas1e** [comment]
Yeah, we found it yesterday, as well, Daniel understand format and foramt is simply : header string, state, bin type, bin size, payload, repeat. The state is basically a number with many bits which uniquely identifies a generated shader prog. Right ?

Anyway, we also do profiling of the NeverBall loading, and after that we can conclude that NOVA's CompileShader() function is crap and writen bad.

I wrote big post on our forum, maybe you will be in interest to read it:

Ok, now, some new problem/question.

In GL4ES ptitSeb add precompiled shader archive support (PSA), which mean that we can use precompiled binaries, and not compile shaders anymore. Daniel add to recent ogles2 necessary stuff to make it works, and so we meet with some problem.

For first, to explain, precompile shaders archive (PSA), mean to collect all the compiled shaders as binaries in some place, which is then on running of programm loaded to the memory, and then used just as glUseProgramm() without needs to compile them => no speed loss. 

In our, amigaos4 case, shader compilation and execution looks like this: 

- gl4es generate a GLSL shader, and send it to ogles2.
- ogles2 translate GLSL shader to SPIRV + add necessary patching and stuff (so by logic heavy things).
- NOVA translate that ready to use SPIRV to assembly machine code of gfx card, which then execute (don't know how heavy is it).

Ideally, PSA binaries should be of that last translation (after NOVA have machine code ready from SPIRV), but as Nova didn't provide any public functions for us to save/get them , we at least trying to get rid of ogles2's step: compiling GLSL to SPIRV + patching.

So, we doing all this, just to understand that all the problems is not sending GLSL shader to ogles2, and not ogles2 doing translation from GLSL to SPIRV + patching (which is heavy thing!), but NOVA's CompileShader() :(

I do 2 profiling tests:

1. without PSA support (so, all steps there, including ogles2's GLSL to SPIRV + patching):

http://kas1e.mikendezign.com/aos4/gl4es/PSA/NOpsa_neverball_profile.txt

As you can see, CompileShader() there take almost the same amount of time, showing that most of time spend in NOVA's version. But that all can be misleading of course, and so,

2. second result WITH PSA support, i.e. binaries placed in the SPIRV format on the disk, loaded to the memory on running, and then ogles2 don't do any CompileShader() thing (as can be seen from the next profiling file), but only UsePRogramm() which take nothing:

http://kas1e.mikendezign.com/aos4/gl4es/PSA/psa_neverball_profile.txt

As can be seen, all the time still catched by the Nova's compileshader(). 

@Hans

Have you any idea why the translation / patching etc. of GLSL only takes a tiny fraction of the time of the Nova's SPIR-V assembly step (by tiny fraction means 5% only). Because i can understand if it was other way around, but not like it now.

The much-much more complex GLSL translation and all the internal source-code-patching in ogles2 only wasts 200 ms there.

Is there anything which can be taken care of ?

Also Nova's CompileShaders does not upload the stuiff to the actuall hardware (at least it should be that way?). It should really be just something like an assembler / low-level-compiler with register mapping which should works more or less fast (at least on level of how ogles2's GLSL->SPIRV+patching compiling happens).

So.. if there is any reall reassons why should Nova's CompilerShader() be that slow and can't be optimized by any of reassons, can you made some public functions, by which we can generate ready to use assembly code to binary, and use them via something like useprogramm() , etc. Yes, those binaries will be different on all machines, and need to be generated on each, but then, only one time when first time plays.

**Comment by kas1e** [comment]
Maybe we of course don't understand the "heavy" part of NOVA's CompilerShader(), but still it should't be that slow.

Did you know on Pandora what internals of CompileShader() ? I mean, if PSA change it for you pretty much visually, that mean that Pandora's CompileShader() not very fast too, and doint not only conversion to assembly ?

**Comment by ptitSeb** [comment]
From my understanding, the heavy phase is not the "Compile" but the "Linking". That's why they created the `glProgramBinary(...)` function, because the existing `glShaderBinary(...)` wasn't enough to save time, as it avoid only the "Compile" of the shader and not the link. So yes, on the Pandora, it's the Link phase that takes time, especially when there is a lot of uniform in the program.

Also, by design, the compile / linking of shaders program doesn't need to be fast. Also the execution of the program need to be fast. The compile and linking is supposed to be done at a loading phase of the game, where performance is not an issue. And most games do that: loading shaders program at loading, like other game assets. Dynamically creating new shader program in-game is not a common practice.

**Comment by kas1e** [comment]
Hi!

Its anyway turned out that NOVA's shader compiler is slow crap :) Hans says that he even have compiler optimisation DISABLED for it, because boost shared pointers somehow lost track of their usage counters with optimization enabled, resulting in objects being destroyed while they're still in use.

So no one know how good his code at all with usage of those shared ptrs (probabaly also not optimised), as well as bug happens when optimisation enabled, mean that there is real bug present which need to be fixed, but he instead just disable optimisation.

Anyway, i promised some little gift when you will add PSA, so i send just 50$ in hope to send another 50$ later when Hans will optimise things on his part, and maybe adding necessary functions for save the binary in machine code .. But that what he say:

```
That's not so easy because more needs to be saved than just the raw shader code. There are various register settings that need to be preserved. So, I'd have to come up with another file format to save the binary.  I'm totally swamped with other work, so I'm unlikely to have time for looking into this, sorry.
```

So.. At least gl4es have PSA now, and ogles2 have necessary functions + we fix aglGetProcAddress because of this , which is good things too ..

Thanks !

**Comment by ptitSeb** [comment]
I'm not surprised he needs to save more then just RAW data. There is the result of the Link I guess, with all the internal setup of the shaders program (and what register are linked to what VA / Uniform and the Varying setup between Vertex and Fragment shaders for example). So yeah, a bit a of work to organise hat in a proprer format...

And ye, gl4es have psa now, and it's working well on the Pandora at least.

And thanks for the gift :)

**Comment by Sisah2** [comment]
Hello, im just trying enable precompiled shader archive on android (openmw) and found that there is check in hardext.c for GL_OES_get_program (result is extension not found). Shouldnt it be GL_OES_get_program_binary?

**Comment by ptitSeb** [comment]
Ah maybe. But it does work with the other name too on my hardware. A double test should be done I guess. If you change the test to "GL_OES_get_program_binary", does it work on your side?

**Comment by ptitSeb** [comment]
The destructor an Android, well, if the Constructor attribute is disabled on android (it is the case to avoid lockup on old android version), the contructor and destructor of the lib need to be called manualy. So probably an issue on openmw side (but the system is supposed to do the cleanup anyway).

**Comment by Sisah2** [comment]
Then it say GL_OES_get_program_binary is supported with 1 supported binary format. I think it will work. Just need to save them differently. I try to save .psa when calling glhint with custom arguments. Hope it will work :)
